bindings/cpu_profiler.cc

```diff
@@ -52,6 +57,171 @@ v8::CpuProfilingLoggingMode GetLoggingMode() {
+    isolate(v8::Isolate::GetCurrent()),
+    heap_stats(v8::HeapStatistics())
+  {
+    uv_timer_init(loop, &timer);
+    timer.data = this;
+  };
+
+  static void callback(uv_timer_t* handle);
+  static void add_listener(uv_timer_t* handle, const char* profile_id, std::function<void(uv_timer_t*, uint64_t, v8::HeapStatistics&)> cb);
+  static void remove_listener(uv_timer_t* handle, const char* profile_id, std::function<void(uv_timer_t*, uint64_t, v8::HeapStatistics&)>& cb);
+};
+
+void HeapStatisticsTimer::callback(uv_timer_t* handle) {
+  HeapStatisticsTimer* self = static_cast<HeapStatisticsTimer*>(handle->data);
+  self->isolate->GetHeapStatistics(&self->heap_stats);
+  uint64_t ts = uv_hrtime();
+
+  for (auto& listener : self->listeners) {
+    listener.second(handle, ts, self->heap_stats);
+  }
+}
+
+void HeapStatisticsTimer::add_listener(uv_timer_t* handle, const char* profile_id, std::function<void(uv_timer_t*, uint64_t, v8::HeapStatistics&)> cb) {
+  auto self = static_cast<HeapStatisticsTimer*>(handle->data);
+  self->listeners.emplace(std::string(profile_id), cb);
+
+  if (self->listeners.size() == 1) {
+    uv_timer_set_repeat(&self->timer, self->period_ms);
+    uv_timer_start(&self->timer, callback, 0, self->period_ms);
+  }
+}
+
+void HeapStatisticsTimer::remove_listener(uv_timer_t* handle, const char* profile_id, std::function<void(uv_timer_t*, uint64_t, v8::HeapStatistics&)>& cb) {
+  auto self = static_cast<HeapStatisticsTimer*>(handle->data);
+  self->listeners.erase(std::string(profile_id));
+
+  if (self->listeners.size() == 0) {
+    uv_timer_stop(&self->timer);
+  }
+};
+
+class Profiler {
+public:
+  std::unordered_map<std::string, SentryProfile*> active_profiles = std::unordered_map<std::string, SentryProfile*>();
+
+  v8::CpuProfiler* cpu_profiler;
+  HeapStatisticsTimer heap_statistics_timer = HeapStatisticsTimer(uv_default_loop());
+
+  explicit Profiler(const napi_env& env, v8::Isolate* isolate) :
+    cpu_profiler(
+      v8::CpuProfiler::New(isolate, kNamingMode, GetLoggingMode())) {
+    napi_add_env_cleanup_hook(env, DeleteInstance, this);
+  }
+
+  static void DeleteInstance(void* data);
+};
+
+void Profiler::DeleteInstance(void* data) {
+  Profiler* profiler = static_cast<Profiler*>(data);
+  profiler->cpu_profiler->Dispose();
+  delete profiler;
+}
+
+class SentryProfile {
+private:
+
+  uint64_t started_at = 0;
+  uint16_t heap_measurement_index = 0;
+  std::vector<uint64_t> heap_stats_ts = std::vector<uint64_t>(300);
+  std::vector<uint64_t> heap_stats_usage = std::vector<uint64_t>(300);
+  std::function<void(uv_timer_t*, uint64_t, v8::HeapStatistics&)> memory_sampler_cb = [this](uv_timer_t* timer, uint64_t ts, v8::HeapStatistics& stats) {
+    heap_stats_ts.insert(heap_stats_ts.begin() + heap_measurement_index, ts - started_at);
+    heap_stats_usage.insert(heap_stats_usage.begin() + heap_measurement_index, static_cast<uint64_t>(stats.used_heap_size()));
+    ++heap_measurement_index;
+    };
```

Comment:
same here, you can remove these assignments here and move them to your initializer on line 159